plutus-core/executables/debugger/Event.hs

```diff
@@ -74,38 +73,46 @@ handleDebuggerEvent driverMailbox bev@(B.VtyEvent ev) = do
             pure ()
         _ -> handleEditorEvent
 handleDebuggerEvent _driverMailbox (B.AppEvent (UpdateClientEvent cekState)) = do
-    let uplcHighlight = do
-            uplcSpan <- uplcAnn <$> case cekState of
-                Computing _ _ _ t -> Just (UPLC.termAnn t)
-                Returning _ ctx _ -> contextAnn ctx
-                _                 -> Nothing
+    let uplcHighlight :: Maybe HighlightSpan = do
+            uplcSpan <- uplcAnn <$> cekStateAnn cekState
             pure HighlightSpan
                 { _hcSLoc = B.Location (srcSpanSLine uplcSpan, srcSpanSCol uplcSpan),
                   _hcELoc = Just $ B.Location (srcSpanELine uplcSpan, srcSpanECol uplcSpan)
                 }
-    modify' $ \st -> case cekState of
-        Computing{} ->
-            st & dsUplcHighlight .~ uplcHighlight
+    let sourceHighlight :: Maybe HighlightSpan = do
+            txSpans <- txAnn <$> cekStateAnn cekState
+            -- FIXME: use some/all spans for highlighting, not just the first one
+            firstTxSpan <- S.lookupMin $ coerce txSpans
+            -- TODO: the HS_FILE supplied from the command line gets highlighted
+            -- The highlighting will not make sense or even break if the user provides
+            -- wrong HS_FILE or if the UPLC originated from multiple HS modules.
+            pure HighlightSpan
+                { _hcSLoc = B.Location (srcSpanSLine firstTxSpan, srcSpanSCol firstTxSpan),
+                  _hcELoc = Just $ B.Location (srcSpanELine firstTxSpan, srcSpanECol firstTxSpan)
```

Comment:
This should be `srcSpanECol firstTxSpan - 1`
GHC's `SrcSpan`'s ending column is one larger than the column of the last character.
[It is adjusted when the SrcSpan is pretty-printed](https://gitlab.haskell.org/ghc/ghc/-/blob/7eac2468a726f217dd97c5e2884f6b552e8ef11d/compiler/GHC/Types/SrcLoc.hs#L728).

Reply:
Thanks!